### PR TITLE
Leverage jail(8) for shared IP networking behaviour

### PR DESCRIPTION
**Pull request by phyber**
As discussed in #758, this PR changes the behaviour of `iocage` to let `jail(8)` take on the heavy lifting of managing shared IP jails.

The main change to enable this behaviour is that `iocage` will now write out a `jail.conf` for each jail it manages under `/var/run`, it will then use these configs when starting and stopping the jails.
This allows users to (ip4 given as examples, but it works the same for ip6):
  - Have jailed shared with the host, managed via `/etc/rc.conf`, `ip4.addr=192.168.1.1`.
  - Have jails share IPs with each other, using the above.
  - Have `iocage` (really `jail(8)`) manage the addition and removal of aliases on jail start/stop, `ip4.addr=lo1|127.0.1.1`, etc.

The old behaviour of `iocage` trying to work out:
  - On start: If it should add the "primary" network interface to an `ip4_addr` that was given simply as `1.2.3.4`.
  - On stop: Which `ip4_addr` aliases it should remove via `ifconfig`.

have been completely removed. It would be possible to support both the old and new behaviour, but it gets quite complicated, and I'm not sure there's a good reason to keep them when FreeBSD can manage it via `jail(8)`.

Another small side effect of these changes was that we fixed a small bug where `iocage` could orphan addresses on interfaces if the `ip4_addr` was changed while a jail was running and the jail was then restarted. Vim backup files `*~` were also added to `.gitignore`.

I have not yet squashed this PR, as if changes are requested my history might be useful, but I will squash it if/when the PR is accepted.
I have also not yet fixed up the documentation, pending review of the work so far.

While testing this I've been managing 6-7 jails with varying combinations of address setups successfully, however, I only use shared IP behaviours so I have not tested `vnet` based jails. I believe they should be working as they were before, but someone else may want to test that.

- [x] Explain the feature
- [x] Read [CONTRIBUTING.md](https://github.com/iocage/iocage/blob/master/CONTRIBUTING.md)
- [ ] Documentation updated
- [ ] Squashed

### CODE REVIEW

**Comment by phyber** [comment]
Will take care of the travis issues tomorrow. 👍

File: iocage_lib/ioc_start.py
Comment (skarekrow) [inline iocage_lib/ioc_start.py]: I think these lines can be squashed down to a single line, as ip4_addrs isn't used anywhere else

```
net.append(f'ip4.addr={ip4_addr}')
```

**Review by skarekrow — CHANGES_REQUESTED** [review]
Looks good! Only potential downside right now is that we should remove these sentinel files on stop, as they're just rewritten anyways on start. We should also check for and remove them during destroy/clean. These may pile up quite a bit.

File: iocage_lib/ioc_start.py
Comment (phyber) [reply iocage_lib/ioc_start.py]: Will put that in with the next changes.

**Comment by phyber** [comment]
Fair comment, I did wonder about deleting them on stop, and this is a good reason to do so.

**Review by skarekrow — DISMISSED** [review]

**Comment by phyber** [comment]
I _think_ the changes to `ioc_destroy.py` means that the jail conf files should also be cleaned via the appropriate things in `ioc_clean.py` as they go via the `destroy_jail` function.

**Comment by skarekrow** [comment]
> I _think_ the changes to `ioc_destroy.py` means that the jail conf files should also be cleaned via the appropriate things in `ioc_clean.py` as they go via the `destroy_jail` function.

Yep that should do it. We should perhaps catch exceptions with the destruction process and pass since it's just a sentinel file, and not something that should hold up destroy/clean

**Review by skarekrow — APPROVED** [review]

**Comment by phyber** [comment]
`OSError`s on `unlink` in both stop and destroy are now wrapped to `pass` on failure. I wasn't sure if you wanted the `su.Popen` call to `jail` wrapped in the destroy too, so I've left that alone for the moment.

**Comment by skarekrow** [comment]
Nope I think those should still be viable errors to see. Since we actually want the jail process removed. Thanks again for this PR!

**Comment by phyber** [comment]
If we're happy here, I'll perform the final squash. It doesn't look like anything in the existing documentation really needs changing.

**Comment by phyber** [comment]
Or, I guess github has that Squash Merge button these days :)

**Comment by skarekrow** [comment]
As long as you are done with this I can actually squash and commit it as is. I don't think any documentation should need changing, since this is all (much needed) plumbing.

**Comment by phyber** [comment]
Yup, I believe I'm done if you're happy with it as is.

**Comment by skarekrow** [comment]
:+1: Thanks!

**Pull request merged**
